The BOOM-bardier barrrel

Fueredoriku opened the pull request:
TODO:
- [x] Model
- [x] Texturing
- [x] Animation
- [x] Remote detonation
- [x] Fire extension early detonation
- [x] Enlarger extension barrel variant
- [x] Trash uplink detonation launching
- [x] Fix or workaround #164 
- [x] Add new fitting crosshairmode
- [x] Secret names and loading tips
- [x] SFX
- [x] Maybe #739

Done:
- Fixed #164 
- Fixed #602 
- Fixed issue where objectpools and their holders persisted in scene hierarchy
- Fixed issue where players would not get ignited by fire mesh projectiles
- Fixed issue where decals would sometime get attached to AiTargetPosition prefab
- Added disabling of crosshair when in spectating mode
- Added object pooling for trail fire
- Adjusted fire particle scaling on fire extension model
- Adjusted stickyprojectiles to spawn at oldposition instead for better visuals

Note:
Dynamites can't be detonated in training grounds until the zooming fix in #879 is merged, but it should work fine in all other situations.

Fueredoriku committed:
Add dynamite barrel prefab, item and basic functions

Fueredoriku committed:
Fix #602, fix #164, add Boom-bardier mesh and animations.

- Fixed issue where objectpools and their holders persited in scene hierarchy
- Fixed issue where players would not get ignited by fire mesh projectiles
- Added object pooling for trail fire
- Adjusted fire particle scaling on fire extension model
- Adjusted stickyprojectiles to spawn at oldposition instead for better visuals

Fueredoriku committed:
Add texture to boom-bardier and remote detonation of dynamites

Fueredoriku committed:
Add explosive barrel variant for enlarger

toberge commented on Assets/Scripts/UI/PlayerHUDController.cs:
```diff
@@ -5,6 +5,8 @@
 using static GunStats;
 using System;
+using Org.BouncyCastle.Asn1.Cmp;
+using Unity.VisualScripting.Antlr3.Runtime.Misc;
```
Dette skulle ikke vært importert?

Fueredoriku replied:
Cursed imports that I have no idea how got there

toberge commented on Assets/Scripts/UI/PlayerHUDController.cs:
```diff
@@ -373,9 +379,27 @@ private void SetHitmarkScale(float scale)
+            LeanTween.cancel(crosshairRadiusTween);
+            crosshairMaterial.SetFloat("_Radius", initialCrosshairRadius);
+        }
+        crosshairRadiusTween = LeanTween.value(crosshair.gameObject, SetCrosshairRadius, initialCrosshairRadius, 0f, 0.2f)
```
This animation was a little too brief, perhaps increase its duration by 100 ms or so?

Fueredoriku replied:
I was afraid the feedback wouldn't be fast enough, but after testing it seems to be ok, so I set it to 0.3 👍

toberge commented on Assets/Scripts/Utils/StuckObject.cs:
```diff
@@ -0,0 +1,12 @@
+
+/// <summary>
+/// Interface component for objectpooling sticky projectiles.
+/// This is necessary because you can't refference GameObject as a Template type.
```
Typo in reference.

Fueredoriku replied:
Fixed

toberge commented on Assets/Scripts/Augment/BulletModifiers/StickyProjectileModifier.cs:
```diff
@@ -35,26 +45,56 @@ public Priority GetPriority()
+    {
+        var stuck = isDespawnedAfterTime ?
+        stuckObjects.GetAndReturnLater(stuckLifeTime)
+        : stuckObjects.Get();
```
Indent >:)

Fueredoriku replied:
🤠

toberge commented on Assets/Scripts/Augment/AugmentImplementations/Hat/MeshProjectileController.cs:
```diff
@@ -34,6 +34,10 @@ public class MeshProjectileController : ProjectileController
     private ProjectileState[] projectiles;
 
+    public Vector3[] ProjectilePositions => 
+        projectiles.Where(p => p != null && p.active == true)
```
```csharp
p is { active = true }
```

Fueredoriku replied:
Our C# version is too dumb 😥

toberge commented on Assets/Scripts/Augment/AugmentImplementations/DynamiteBarrel.cs:
```diff
@@ -0,0 +1,137 @@
+        }
+        else if (sateliteUplink)
+            sateliteUplink.TargetManual(activeDynamites
+                .Select(dynamite => dynamite.transform.position).ToArray());
```
No need to pass an array, change type to IEnumerable and it accepts the linq query just fine

Fueredoriku replied:
Agree, only reason I didn't do this was because I was looking for a clean alternative to passing a collection into the method at all, but I didn't clean up after giving up lol

toberge commented on Assets/Scripts/Augment/AugmentImplementations/DynamiteBarrel.cs:
```diff
@@ -0,0 +1,137 @@
+            return;
+        }
+        syncDirection = SyncDirection.ClientToServer;
+        gunController.Player.GetComponent<PlayerMovement>().ResetZoom();
```
Please avoid calling ResetZoom etc without checking if this player has an input.... they may be network players

Fueredoriku replied:
We should really refactor the input mangager subscriptions sheningans at some point

Fueredoriku committed:
Add Boombardier SFX, crosshair, extenstion combos, add #739

- Add early detonation with fire extension
- Add explosive barrel variant with enlarger extension
- Add trashuplink detonation targeting
- Remove crosshairs when spectating
- Fix issue where decals could be placed on AiTargetPosition prefab
- Add explosion screen shakes for closing #739
- Add new secret names and loading screen tips

toberge reviewed (approved):
is prolly fine
